unixfs/src/dir/walk.rs

```diff
@@ -0,0 +1,777 @@
+            ref x => todo!("dir after {:?}", x),
+        }
+    }
+
+    fn as_bucket(
+        &mut self,
+        cid: Cid,
+        name: &str,
+        depth: usize
+    ) {
+        use InnerKind::*;
+        match self.kind {
+            RootBucket | Bucket(_) | File(_, None) | Symlink(_) => {
+                self.kind = Bucket(cid);
+
+                if name.is_empty() {
+                    // continuation bucket going bucket -> bucket
+                    while self.depth > depth {
+                        assert!(self.path.pop());
+                        self.depth -= 1;
+                    }
+                } else {
+                    self.set_path(name, depth);
+                }
+
+                assert_eq!(self.depth, depth, "{:?}", self.path);
+            },
+            ref x => todo!("bucket after {:?}", x),
+        }
+    }
+
+    fn as_file(
+        &mut self,
+        cid: Cid,
+        name: &str,
+        depth: usize,
+        metadata: FileMetadata,
+        step: Option<FileVisit>,
+    ) {
+        use InnerKind::*;
+        match self.kind {
+            RootDirectory
+            | RootBucket => {
+                assert_eq!(self.depth, 0);
+                self.kind = File(cid, step);
+                self.set_path(name, depth);
+                self.metadata = metadata;
+            },
+            Bucket(_)
+            | Directory(_)
+            | File(_, None)
+            | Symlink(_) => {
+                self.kind = File(cid, step);
+                self.set_path(name, depth);
+                self.metadata = metadata;
+            }
+            ref x => todo!("file from {:?}", x),
+        }
+    }
+
+    fn as_symlink(
+        &mut self,
+        cid: Cid,
+        name: &str,
+        depth: usize,
+        metadata: FileMetadata
+    ) {
+        use InnerKind::*;
+        match self.kind {
+            RootDirectory => {
+                self.kind = Symlink(cid);
+                self.set_path(name, depth);
+                self.metadata = metadata;
+            },
+            Bucket(_)
+            | Directory(_)
+            | File(_, None)
+            | Symlink(_) => {
+                self.kind = Symlink(cid);
+                self.set_path(name, depth);
+                self.metadata = metadata;
+            },
+            ref x => todo!("symlink from {:?}", x),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Walk<'a> {
+    /// Walk was started on a file node which we don't know any name for
+    SingleBlockFile {
+        bytes: &'a [u8],
+        metadata: FileMetadata,
+    },
+    Symlink {
+        metadata: FileMetadata,
+        target: &'a [u8],
+    },
+    /// Walk was started on a file which can be iterated further with the given FileVisit
+    // TODO: not sure why not expose this through the continue_block as well? the path will just be
+    // empty all the time.
+    MultiBlockFile {
+        metadata: FileMetadata,
+        visit: FileVisit,
+    },
```

Comment:
alternative would to just add the `Option<FileVisit>` to `Walk::SingleBlockFile` and rename it as `File`.